lib/charms/layer/openstack.py

```diff
@@ -32,52 +32,58 @@ def get_credentials():
     """
     config = hookenv.config()
 
-    # try credentials config
-    if config['credentials']:
-        try:
-            creds_data = b64decode(config['credentials']).decode('utf8')
-            creds_data = json.loads(creds_data)
-            log('Using "credentials" config values for credentials')
-            _save_creds(creds_data)
-            return True
-        except Exception:
-            status.blocked('invalid value for credentials config')
-            return False
-    no_creds_msg = 'missing credentials; set credentials config'
-
-    # try individual config
-    if any([config['auth-url'],
-            config['username'],
-            config['password'],
-            config['project-name'],
-            config['user-domain-name'],
-            config['project-domain-name'],
-            config['region']]):
-        log('Using individual config values for credentials')
-        _save_creds(config)
-        return True
+    required_fields = [
+        'auth_url',
+        'region',
+        'username',
+        'password',
+        'user_domain_name',
+        'project_domain_name',
+        'project_name',
+    ]
+    optional_fields = [
+        'endpoint_tls_ca',
+    ]
+    # pre-populate with empty values to avoid key and arg errors
+    creds_data = {field: '' for field in required_fields + optional_fields}
 
     # try to use Juju's trust feature
     try:
+        log('Checking credentials-get for credentials')
         result = subprocess.run(['credential-get'],
                                 check=True,
                                 stdout=subprocess.PIPE,
                                 stderr=subprocess.PIPE)
-        creds_data = yaml.load(result.stdout.decode('utf8'))
-
-        log('Using credentials-get for credentials')
-        _save_creds(creds_data)
-        return True
+        _creds_data = yaml.load(result.stdout.decode('utf8'))
+        _merge_if_set(creds_data, _normalize_creds(_creds_data))
     except FileNotFoundError:
         pass  # juju trust not available
     except subprocess.CalledProcessError as e:
         if 'permission denied' not in e.stderr.decode('utf8'):
             raise
-        no_creds_msg = 'missing credentials access; grant with: juju trust'
 
-    # no creds provided
-    status.blocked(no_creds_msg)
-    return False
+    # merge in combined credentials config
+    if config['credentials']:
+        try:
+            log('Using "credentials" config values for credentials')
+            _creds_data = b64decode(config['credentials']).decode('utf8')
+            _creds_data = json.loads(_creds_data)
+            _merge_if_set(creds_data, _normalize_creds(_creds_data))
+        except Exception:
+            status.blocked('invalid value for credentials config')
+            return False
+
+    # merge in individual config
+    _merge_if_set(creds_data, _normalize_creds(config))
```

Comment:
will this add all configuration options to creds_data? It looks like it will just dump anything with a value into it. Is that desired?

Reply:
nevermind, _normalize_creds...